Assignments/ip2/ip2.md

```diff
@@ -0,0 +1,310 @@
+
+### Task 3: Implement the Tic-Tac-Toe Board (16 points total)
+This task is to implement the `TicTacToeBoard` component, which will render the actual (interactive) board. It is located in the file `frontend/src/components/Town/interactables/TicTacToe/TicTacToeBoard.tsx`.
+
+<details><summary markdown="span">View the specification for this component</summary>
+{% highlight typescript %}
+/**
+ * A component that renders the TicTacToe board
+ *
+ * Renders the TicTacToe board as a "StyledTicTacToeBoard", which consists of 9 "StyledTicTacToeSquare"s
+ * (one for each cell in the board, starting from the top left and going left to right, top to bottom).
+ * Each StyledTicTacToeSquare has an aria-label property that describes the cell's position in the board,
+ * formatted as `Cell ${rowIndex},${colIndex}`.
+ *
+ * The board is re-rendered whenever the board changes, and each cell is re-rendered whenever the value
+ * of that cell changes.
+ *
+ * If the current player is in the game, then each StyledTicTacToeSquare is clickable, and clicking
+ * on it will make a move in that cell. If there is an error making the move, then a toast will be
+ * displayed with the error message as the description of the toast. If it is not the current player's
+ * turn, then the StyledTicTacToeSquare will be disabled.
+ *
+ * @param gameAreaController the controller for the TicTacToe game
+ */
+export default function TicTacToeBoard({ gameAreaController }: TicTacToeGameProps): JSX.Element
+{% endhighlight %}
+</details>
+
+There is significant ambiguity in the specification when it comes to exactly how this looks. We will not evaluate your submission on the basis of how closely it looks like our referencence implementation: it need only be functionally correct (as defined by the included test cases).
+
+Grading for implementation tasks:
+* Drawing the board for observers: 15 points
+* Drawing the interactive board for players: 25 points
+
+All of the tests are provided in the handout. Run the tests for this task by running the command `npx jest TicTacToeBoard` in the `frontend` directory (for convenience, you may want to use `npx jest --watch` ...). You will also likely find it convenient to run the app in your browser while you work on this task for interactive debugging.
+
+The grading script will assign full marks for each implementation task if all of the tests for that task pass.  There is no partial credit. 
+
+### Task 4: Implement the Leaderboard (10 points total)
+
+This task is to implement the `Leaderboard` component, which will render a list of `GameResult`s as a leaderboard. It is located in the file `frontend/src/components/Town/interactables/Leaderboard.tsx`.
+
+<details><summary markdown="span">View the specification for this component</summary>
+{% highlight typescript %}
+
+/**
+ * A component that renders a list of GameResult's as a leaderboard, formatted as a table with the following columns:
```

Comment:
With respect to this spec, my TA says:

> I noticed that the specification expects the leaderboard to be sorted from most wins to least,
> but the test that checks for this case may pass even if the implementation does not explicitly
> sort the leaderboard because the test case’s sorted order is in the same order that the players
> are introduced in the results array. (If you comment out the line that sorts the leaderboard in
> the reference solution, the tests still pass)

Reply:
Good find! I really wish that there were a more effective way to run mutation testing on react tests... Fixed.